File: .tmp-lot18/mesure2.mjs
import { RIB_LETTERS, checkBankDetails } from '../content/snippets/check-bank-details-before-a-transfer/n0.js';
const SOSIES = { 0: 'O', 1: 'I', 2: 'Z', 5: 'S', 6: 'G', 8: 'B' };
const cleIso = (pays, bban) => String(98n - (BigInt([...(bban + pays + '00')].map((c) => parseInt(c, 36)).join('')) % 97n)).padStart(2, '0');
const cleRib = (b, g, c) => String((97n - (BigInt(`${[...(b + g + c)].map((x) => RIB_LETTERS[x] ?? x).join('')}00`) % 97n)) % 97n).padStart(2, '0');
function alea(graine) { let etat = graine >>> 0; return () => { etat = (etat * 1664525 + 1013904223) >>> 0; return etat / 2 ** 32; }; }
function ibansFrancais(nombre, graine = 3) {
  const suivant = alea(graine); const entier = (max) => Math.floor(suivant() * max); const sortie = [];
  while (sortie.length < nombre) {
    const banque = String(10000 + entier(90000));
    const guichet = String(entier(100000)).padStart(5, '0');
    const compte = String(entier(10 ** 11)).padStart(11, '0');
    const bban = banque + guichet + compte + cleRib(banque, guichet, compte);
    const candidat = `FR${cleIso('FR', bban)}${bban}`;
    if (checkBankDetails(candidat).valid) sortie.push(candidat);
  }
  return sortie;
}
let ensemble = 0, seuleIso = 0, passees = 0;
for (const numero of ibansFrancais(500)) {
  for (let rang = 4; rang < numero.length; rang += 1) {
    if (!(numero[rang] in SOSIES)) continue;
    ensemble += 1;
    const faute = numero.slice(0, rang) + SOSIES[numero[rang]] + numero.slice(rang + 1);
    const rapport = checkBankDetails(faute);
    if (rapport.valid || rapport.national_key === false) seuleIso += 1;
    if (rapport.valid) passees += 1;
  }
}
console.log('ensemble', ensemble, 'seuleIso', seuleIso, 'passees', passees, 'taux', (100 * seuleIso / ensemble).toFixed(4) + '%');
let e=0,p=0;
for (const numero of ibansFrancais(100)) {
  for (let rang = 4; rang < numero.length; rang += 1) {
    for (const chiffre of '0123456789') {
      if (chiffre === numero[rang]) continue;
      e += 1;
      if (checkBankDetails(numero.slice(0,rang)+chiffre+numero.slice(rang+1)).valid) p += 1;
    }
  }
}
console.log('un-chiffre essais', e, 'passees', p);
let e2=0,p2=0;
for (const numero of ibansFrancais(100)) {
  for (let rang = 4; rang < numero.length - 1; rang += 1) {
    if (numero[rang] === numero[rang+1]) continue;
    e2 += 1;
    const permute = numero.slice(0,rang)+numero[rang+1]+numero[rang]+numero.slice(rang+2);
    if (checkBankDetails(permute).valid) p2 += 1;
  }
}
console.log('transpositions essais', e2, 'passees', p2);

File: .tmp-lot18/mesure2.py
import sys, random
sys.path.insert(0, 'content/snippets/check-bank-details-before-a-transfer')
from stdnum import iban as registre
from n0 import RIB_LETTERS, check_bank_details

SOSIES = {"0": "O", "1": "I", "2": "Z", "5": "S", "6": "G", "8": "B"}

def cle_iso(pays, bban):
    etendu = "".join(str(int(c, 36)) for c in bban + pays + "00")
    return "%02d" % (98 - int(etendu) % 97)

def cle_rib(banque, guichet, compte):
    chiffres = "".join(RIB_LETTERS.get(c, c) for c in banque + guichet + compte)
    return "%02d" % ((97 - int(chiffres + "00") % 97) % 97)

def ibans_francais(nombre, graine=3):
    alea = random.Random(graine)
    sortie = []
    while len(sortie) < nombre:
        banque, guichet = "%05d" % alea.randint(10000, 99999), "%05d" % alea.randint(0, 99999)
        compte = "%011d" % alea.randint(0, 10**11 - 1)
        bban = banque + guichet + compte + cle_rib(banque, guichet, compte)
        candidat = "FR" + cle_iso("FR", bban) + bban
        if registre.is_valid(candidat):
            sortie.append(candidat)
    return sortie

seule_iso = ensemble = passees = 0
for numero in ibans_francais(500):
    for rang in range(4, len(numero)):
        if numero[rang] not in SOSIES:
            continue
        ensemble += 1
        faute = numero[:rang] + SOSIES[numero[rang]] + numero[rang + 1:]
        if registre.is_valid(faute):
            seule_iso += 1
        if check_bank_details(faute)["valid"]:
            passees += 1
print('ensemble', ensemble, 'seule_iso', seule_iso, 'passees', passees,
      'taux %.4f%%' % (100 * seule_iso / ensemble))
